api/src/context.rs

```diff
@@ -797,27 +806,168 @@ impl Context {
+                    }
+                },
+                Err(_) => {
+                    break;
+                },
+            }
+        }
+        if blocks.is_empty() && !cached_blocks_hit {
+            let estimation = self.default_gas_estimation(min_gas_unit_price);
+            cache.estimation = Some(estimation);
+            cache.last_updated_epoch = Some(epoch);
+            cache.last_updated_time = Some(Instant::now());
+            return Ok(estimation);
+        }
+        let remaining = max_block_history - blocks.len();
+
+        // 2. Get gas prices per block
+        let mut min_inclusion_prices = vec![];
+        let full_block_txns = self.node_config.api.gas_estimation_full_block_txns;
+        // TODO: if multiple calls to db is a perf issue, combine into a single call and then split
+        for (first, last) in blocks {
+            let min_inclusion_price =
+                match self
+                    .db
+                    .get_gas_prices(first, last - first, ledger_info.ledger_version.0)
+                {
+                    Ok(prices) => {
+                        if prices.len() < full_block_txns {
+                            min_gas_unit_price
+                        } else {
+                            prices.iter().min().unwrap() + 1
+                        }
+                    },
+                    Err(_) => min_gas_unit_price,
+                };
+            min_inclusion_prices.push(min_inclusion_price);
+            cache
+                .min_inclusion_prices
+                .insert((epoch, last), min_inclusion_price);
+        }
+        if cached_blocks_hit {
+            for (_, v) in cache
+                .min_inclusion_prices
+                .range((Included(&(epoch, 0)), Included(&(epoch, lookup_version))))
+                .rev()
+                .take(remaining)
+            {
+                min_inclusion_prices.push(*v);
+            }
+        }
+
+        // 3. Get values
+        // (1) low
+        let low_price = match min_inclusion_prices.iter().take(10).min() {
+            Some(price) => *price,
+            None => min_gas_unit_price,
+        };
+
+        // (2) market
+        let mut latest_prices: Vec<_> = min_inclusion_prices.iter().take(30).cloned().collect();
+        latest_prices.sort();
+        let market_price = latest_prices[latest_prices.len() / 2];
+
+        // (3) aggressive
+        min_inclusion_prices.sort();
+        let p90_price = min_inclusion_prices[min_inclusion_prices.len() * 9 / 10];
+        // round up to next bucket
+        let aggressive_price = self.next_bucket(p90_price);
+
+        let estimation = GasEstimation {
+            deprioritized_gas_estimate: Some(low_price),
+            gas_estimate: market_price,
+            prioritized_gas_estimate: Some(aggressive_price),
+        };
+        // 4. Update cache
+        // GC old entries
+        if cache.min_inclusion_prices.len() > max_block_history {
```

Comment:
Do we want to add a unit test to ensure we are doing garbage collection, i.e., that the cache never grows indefinitely?